### PR TITLE
[Android] Adding prop android_minWidth to Switch component

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary

This issue fixes https://github.com/facebook/react-native/issues/29047
Adding prop android_minWidth to Switch component.

## Changelog

[Android] [Fixed] - Adding prop android_minWidth to Switch component

## Test Plan

**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

I will write separate pull request to update the docs on https://github.com/facebook/react-native-website

| **AFTER** |
|:-------------------------:|
|  |

</p>
</details>

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java [hunk 1]
```diff
@@ -175,6 +175,12 @@ public void setTrackTintColor(ReactSwitch view, @Nullable Integer color) {
     view.setTrackColor(color);
   }
 
+  @ReactProp(name = "androidMinWidth")
+  public void setSwitchMinWidth(ReactSwitch view, Integer pixels) {
+    Integer width = pixels == null ? view.getSwitchMinWidth() : pixels;
+    view.setSwitchMinWidth(width);
```
Comment (HectorRicardo) [inline ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java:181]: I just have a comment (or question). React-Native uses dip (density independent pixels) for the width and size of the components., as stated in the docs: [https://reactnative.dev/docs/height-and-width](https://reactnative.dev/docs/height-and-width).

Shouldn't this method expect a dip value instead of pixels ? The conversion from dip to pixels can be handled using this method:

```java
  private float dpToPixels(float dp) {
    return TypedValue.applyDimension(
      TypedValue.COMPLEX_UNIT_DIP,
      dp,
      view.getContext().getResources().getDisplayMetrics()
    );
  }
```

**Comment by HectorRicardo** [comment]
Hey Fabrizio, thanks for the pull request!

I added a comment to one of the files, suggesting the use of dip (density independent pixels), instead of pixels.

File: ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java (hunk 1 above)
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java]: @HectorRicardo 

I added https://github.com/facebook/react-native/pull/29059/commits/7c6901e72b511565bd2ce2c1b6d91de2ef2f1956 to use 

I quote the [dpi definition](https://en.wikipedia.org/wiki/Device-independent_pixel)

>A device-independent pixel (also: density-independent pixel, dip, dp) is a unit of length.

>For example, on the Android operating system a device-independent pixel is equivalent to one physical pixel on a 160 dpi screen,[1] while the Windows Presentation Foundation specifies one device-independent pixel as equivalent to 1/96th of an inch.[2]

>As **dp** is a physical unit it has an absolute value which can be measured in traditional units, e.g. for Android devices 1 dp equals 1/160 of inch or 0.15875 mm.

Increasing `density` of pixels does not change size

image

`setSwitchMinWidth` specifies the `MINUMUM WIDTH`. The actual width will be specified with the [`style` prop](https://reactnative.dev/docs/view-style-props). The `style` prop will not be applied if the `minWidth` is not respected.


| **Density 300 - switch width 50 minWidth 50** | **Density 300 - switch width 50 minWidth 100** |
|:-------------------------:|:-------------------------:|
|  |  |

Persisting Issues with the actual width of this component could be related to https://github.com/facebook/react-native/issues/28235
In that case I would write a separate pull request to fix that dpi issue and fix the following PixelUtil or the View Style prop.

https://github.com/facebook/react-native/blob/851d01b0aa45285bd6a52c89ab9c1b64a637258e/ReactAndroid/src/main/java/com/facebook/react/uimanager/PixelUtil.java#L13-L20

| **Density 200 - switch is 50px width** | **Density 200 - switch is 200px width** |
|:-------------------------:|:-------------------------:|
|  |  |

| **Density 300 - switch is 50px width** | **Density 300 - switch is 200px width** |
|:-------------------------:|:-------------------------:|
|  |  |

**Comment by dulmandakh** [comment]
why not support minWidth in style?

File: ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java (hunk 1 above)
Comment (HectorRicardo) [reply ReactAndroid/src/main/java/com/facebook/react/views/switchview/ReactSwitchManager.java]: Thanks for the explanation Fabrizio! It makes total sense what you're saying, that the style's width will not be applied if min width is not respected.

**Comment by HectorRicardo** [comment]
@dulmandakh that could also work, but I guess it is because the minimum width is only supported on Android and not on iOS, so by naming the property `androidMinWidth` it is clear that it only applies to android

**Comment by fabriziobertoglio1987** [comment]
Thanks for the support @HectorRicardo 

@dulmandakh 

Thanks and Sorry, but I do not work on iOS. I can not tell if this is possible or not.
If you request, I will move this logic to the style prop.

**TLDR**
**I believe It is possible** to build this logic on iOS. 
The iOS Contributor would have to set a check on the `width` style prop on `iOS` side and over-ride that value based on `minWidth` value.

I did a brief research, this are my **inconclusive** results on iOS:

----------

on iOS we are using [UISwitch](https://developer.apple.com/documentation/uikit/uiswitch)

https://github.com/facebook/react-native/blob/a50fa552a7e07f611d626fae99a6ef80845d9d4d/React/Fabric/Mounting/ComponentViews/Switch/RCTSwitchComponentView.mm#L22-L24

as explained [on stackoverflow](https://stackoverflow.com/a/25106983/11452286) 

`UISwitch` inherits from `UIView`. The width/height could be changed using [`UIView transform`][1] 

```c
UISwitch *aSwitch = [[UISwitch alloc] initWithFrame:CGRectMake(120, 120, 51, 31)];
aSwitch.transform = CGAffineTransformMakeScale(2.0, 2.0);
```

Seems difficult to change the width without using [`trasforms`][2] as said [here][3]

image

I personally would not use `transform` for the width/height prop. I would use `transform` for the [transforms][2]. I would not give the user a workaround to set `UISwitch` width/height on iOS using transforms, as this may not work as he expects. 

The iOS developer would have to find a way to set the `UISwitch` width/height and then change the logic to take in consideration the `minWidth` value.

I could not find existing iOS api to set the minWidth of a UISwitch, I did not make extensive research as I want to specialize on Android. 

Thanks a lot.
I wish you a good day

[1]: https://developer.apple.com/documentation/uikit/uiview/1622459-transform
[2]: https://reactnative.dev/docs/transforms
[3]: https://stackoverflow.com/a/25104771/11452286

**Comment by HectorRicardo** [comment]
I also did some research, and although I don't have any Apple device to test, I am pretty sure the default UISwitch is not customizable. I suggest we leave the androidMinWidth attribute outside of the style object.

**Comment by fabriziobertoglio1987** [comment]
>why not support minWidth in style?

Thanks a lot @dulmandakh for your code review. 

The main reason is that I am specializing on Android, Java and ReactNative.
Currently I don't work with ObjectiveC and iOS. 

It is harder for me to work with many different languages and tools.

Thanks a lot
